### PR TITLE
feat: add options.experimental_output

### LINKED ISSUE #50: content_scripts defined in manifest.json won't update after HMR without a full reload of the web extension

### PR DESCRIPTION
**Pull request by Jack-Works**
close #50

### options.experimental_output

**This is an experimental API. API might change at any time. Please provide feedback!**

##### string

If you don't strictly rely on [run_at](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts#run_at),
set it as the following

```js
export default {
  entry: {
    myContentScript: 'src/contentScript.ts',
  },
  // ...
  plugins: [
    // ...
    new WebExtensionPlugin({
      // ...
      experimental_output: {
        myContentScript: 'cs.js'
      },
    })
  ]
}
```

```jsonc
{
  // ...
  "content_scripts": [
    {
      "matches": ["..."],
      "js": ["cs.js"]
    }
  ]
}
```

##### function

If you cannot use asynchronous loading, set up like below.
This setup requires you to have a `manifest.json` being emitted.

```js
export default {
  entry: {
    myContentScript: 'src/contentScript.ts',
  },
  // ...
  plugins: [
    // ...
    new WebExtensionPlugin({
      // ...
      experimental_output: {
        myContentScript: (manifest, list) => {
          manifest.content_scripts[0].js = list
        }
      },
    })
  ]
}
```

##### object

```js
export default {
  entry: {
    background: 'src/contentScript.ts',
  },
  // ...
  plugins: [
    // ...
    new WebExtensionPlugin({
      // ...
      experimental_output: {
        background: {
          file: 'sw.js',
          touch(manifest, file) {
            manifest.background.service_worker = file
          }
        },
      },
    })
  ]
}
```

#### Explanation

**This is an experimental API.**
**API might change at any time.**
**Please provide feedback!**

This option helps the initial chunk loading of content scripts/the background service worker,
usually needed when `optimization.runtimeChunk` or `optimization.splitChunks.chunks` is used.

This option accepts an object, where the keys are the entry name,
and the value is described below.

This option replaces the HTMLWebpackPlugin where the background service worker and content scripts
do not use HTML to load files.

If the value is a `string` (an output file name), for content scripts, it creates an extra
entry file to load all initial chunks **asynchronously** via dynamic import.
This asynchronous loading behavior is limited to the platform limit and **breaks**
[run_at](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts#run_at).

If the value is a `string` (an output file name), for the background service worker (specified
via `options.background.serviceWorkerEntry`), it creates an extra entry file to load all
initial chunks **synchronously**.

The file name specified MUST NOT be any existing file.

If the value is a `function` (`(manifest: any, chunks: string[]) => void`), it requires
a "manifest.json" in the emitted files and lets you edit it on the fly to include all
the initial chunks. This option does not apply to the background service worker because
`manifest.json` does not accept multiple files.

If the value is an `object` (`{ file: string; touch(manifest: any, file: string): void }`),
it generates a new file (see the behavior of `string` above) and provides a callback to
edit the `manifest.json` (see the behavior of `function` above).

If the value is `false`, it asserts that this entry does not have more than one initial file,
otherwise, it will be a compile error.

If the value is `undefined`, it silences the warning for the background service worker.

You can also change your configuration to avoid `optimization.runtimeChunk` or `optimization.splitChunks.chunks`,
in this case, webpack only generates 1 initial file so you don't need this option.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -245,10 +245,104 @@ If you experienced a compatibility issue with any of the following plugins, plea
...
 
+### options.contentScript.experimental_output
+
+**This is an experimental API. API might change at any time. Please provide feedback!**
+
+#### TLDR: How to use this
+
+##### string
+
+If you don't strictly rely on [run_at](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts#run_at),
```
Comment (cezaraugusto) [inline README.md:256]: what would happen for cases where users have `run_at` defined?

File: README.md [hunk 2]
```diff
@@ -245,10 +245,104 @@ If you experienced a compatibility issue with any of the following plugins, plea
...
+    myContentScript: 'src/contentScript.ts',
+  },
+  // ...
+  plugins: [
+    // ...
+    new WebExtensionPlugin({
+      // ...
+      contentScript: {
+        experimental_output: {
+          myContentScript: 'cs.js'
```
Comment (cezaraugusto) [inline README.md:271]: so `cs.js` refers to the output file for content_scripts within the root of the output path? like `dist/cs.js`?

File: README.md [hunk 3]
```diff
@@ -245,10 +245,104 @@ If you experienced a compatibility issue with any of the following plugins, plea
...
+  },
+  // ...
+  plugins: [
+    // ...
+    new WebExtensionPlugin({
+      // ...
+      contentScript: {
+        experimental_output: {
+          myContentScript: (manifest, list) => {
+            manifest.content_scripts[0].js = list
```
Comment (cezaraugusto) [inline README.md:309]: could this handle more than one content_script defined in the manifest file? like which list would this be if I have a `manifest.content_scripts[1].js`?

File: README.md (hunk 1 above)
Comment (Jack-Works) [reply README.md:256]: `cs.js` is like this:

```js
;(async () => {
	const getURL = typeof browser === "object" ? browser.runtime.getURL : chrome.runtime.getURL;
	await import(getURL("runtime.js"));
	await import(getURL("test_txt-util_js.js"));
	await import(getURL("content.js"));
})();
null;
```

therefore it _will not_ run at the very beginning (which may be a hard limit for some content scripts)

File: README.md (hunk 2 above)
Comment (Jack-Works) [reply README.md:271]: Yes, it will be an extra output file.

File: README.md (hunk 3 above)
Comment (Jack-Works) [reply README.md:309]: Yes, this is a function, so you can write any code to change your `manifest.json` (if you defined it via CopyPlugin or other webpack plugins).

```js
{
myContentScript: (manifest, list) => {
    manifest.content_scripts[0].js = list
    manifest.content_scripts[1].js = list
},

myContentScript2: (manifest, list) => {
    manifest.content_scripts[3].js = list
},
}
```

**Comment by Jack-Works** [comment]
I have updated the PR to expand the range to the background service worker. The API is not intuitive enough and I have no confidence with it, so I'll keep it experimental.

**Comment by Jack-Works** [comment]
2.1.0 released, you can have a try!

**Comment by cezaraugusto** [comment]
you're a legend!

**Pull request merged**